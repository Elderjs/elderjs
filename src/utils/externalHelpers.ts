import path from 'path';
import fs from 'fs-extra';

import { QueryOptions, SettingsOptions, Helpers } from './types.js';

let userHelpers;

let cache;
let cacheReloadHash;

async function externalHelpers({
  settings,
  query,
  helpers,
}: {
  settings: SettingsOptions;
  query: QueryOptions;
  helpers: Helpers;
}) {
  const srcHelpers = path.join(settings.srcDir, `helpers/index.js`);
  try {
    if (cache && cacheReloadHash === settings.$$internal.reloadHash) {
      userHelpers = cache;
    } else {
      if (fs.existsSync(srcHelpers)) {
        try {
          const reqHelpers = await import(`${srcHelpers}?hash=${settings.$$internal.reloadHash}`);
          userHelpers = reqHelpers.default || reqHelpers;
          userHelpers = userHelpers.default || userHelpers;

          if (typeof userHelpers === 'function') {
            userHelpers = await userHelpers({ settings, query, helpers });
          }

          if (userHelpers.default) {
            userHelpers = userHelpers.default;
          }

          cacheReloadHash = settings.$$internal.reloadHash;
          cache = userHelpers;
        } catch (err) {
          console.error(err);
        }
      }
    }
  } catch (e) {
    console.error(`Error importing ${srcHelpers}`);
    console.error(e);
  }

  return userHelpers;
}

export default externalHelpers;
