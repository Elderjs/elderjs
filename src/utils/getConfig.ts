import { cosmiconfigSync } from 'cosmiconfig';
import defaultsDeep from 'lodash.defaultsdeep';
import path from 'path';
import fs from 'fs-extra';
import { SettingsOptions, InitializationOptions } from './types.js';
import { getDefaultConfig } from './validations.js';
import prepareFindSvelteComponent from '../partialHydration/prepareFindSvelteComponent.js';
import normalizePrefix from './normalizePrefix.js';
import notProduction from './notProduction.js';
import getFilesAndWatcher from '../core/getFilesAndWatcher.js';

type TCheckCssFiles = {
  cssFiles: string[];
  assetPath: string;
  config: Partial<SettingsOptions>;
  serverPrefix: string;
};
export function getCssFile({ cssFiles, serverPrefix, config, assetPath }: TCheckCssFiles): string {
  if (cssFiles.length > 1) {
    throw new Error(
      `${config.$$internal.logPrefix} Race condition has caused multiple css files in ${assetPath}. If you keep seeing this delete the _elder and ___ELDER___  folders.`,
    );
  }
  if (!cssFiles[0]) {
    throw new Error(`CSS file not found in ${assetPath}`);
  }
  return `${serverPrefix}/_elderjs/assets/${cssFiles[0]}`;
}

function getConfig(initializationOptions: InitializationOptions = {}): SettingsOptions {
  let loadedConfig: InitializationOptions = {};
  const explorerSync = cosmiconfigSync('elder');
  const explorerSearch = explorerSync.search();
  if (explorerSearch && explorerSearch.config) {
    loadedConfig = explorerSearch.config;
  }
  const config: SettingsOptions = defaultsDeep(initializationOptions, loadedConfig, getDefaultConfig());

  const serverPrefix = normalizePrefix(config.prefix || (config.server && config.server.prefix));

  const rootDir = config.rootDir === 'process.cwd()' ? process.cwd() : path.resolve(config.rootDir);
  config.rootDir = rootDir;
  config.srcDir = path.resolve(rootDir, `./${config.srcDir}`);
  config.distDir = path.resolve(rootDir, `./${config.distDir}`);

  const ejsPkg = path.resolve(`./node_modules/@elderjs/elderjs/package.json`);
  let pkgJson;
  if (fs.existsSync(ejsPkg)) {
    pkgJson = fs.readJSONSync(ejsPkg);
  } else {
    pkgJson = { version: 'unknown' };
  }

  config.version = pkgJson.version.includes('-') ? pkgJson.version.split('-')[0] : pkgJson.version;

  config.context = typeof initializationOptions.context !== 'undefined' ? initializationOptions.context : 'unknown';

  config.server = initializationOptions.context === 'server' && config.server;
  config.build = initializationOptions.context === 'build' && config.build;
  config.worker = !!initializationOptions.worker;
  config.prefix = serverPrefix;
  if (serverPrefix && config.server) {
    config.server.prefix = serverPrefix;
  }

  const ssrComponents = path.resolve(config.rootDir, './___ELDER___/compiled/');
  const clientComponents = path.resolve(config.distDir, `.${serverPrefix}/_elderjs/svelte/`);
  const distElder = path.resolve(config.distDir, `.${serverPrefix}/_elderjs/`);
  fs.ensureDirSync(path.resolve(distElder));
  fs.ensureDirSync(path.resolve(clientComponents));

  const { server, client, watcher } = getFilesAndWatcher({
    ...config,
    production: !notProduction(),
    clientComponents,
    ssrComponents,
  });

  config.$$internal = {
    production: !notProduction(),
    ssrComponents,
    clientComponents,
    distElder,
    logPrefix: `[Elder.js]:`,
    serverPrefix,
    findComponent: prepareFindSvelteComponent({
      clientComponents: client,
      ssrComponents: server,
      rootDir,
      srcDir: config.srcDir,
    }),
    files: {
      server,
      client,
    },
    watcher,
  };

  if ((config.css === 'file' || config.css === 'lazy') && initializationOptions.context !== 'test') {
    const assetPath = path.resolve(distElder, `.${path.sep}assets`);
    fs.ensureDirSync(path.resolve(assetPath));
    const cssFiles = fs.readdirSync(assetPath).filter((f) => f.endsWith('.css'));
    config.$$internal.publicCssFile = getCssFile({ cssFiles, serverPrefix, config, assetPath });
  }

  if (config.origin === '' || config.origin === 'https://example.com') {
    console.error(
      `WARN: Remember to put a valid "origin" in your elder.config.js. The URL of your site's root, without a trailing slash. This is frequently used by plugins and leaving it blank can cause SEO headaches.`,
    );
  }

  return config;
}

export default getConfig;
