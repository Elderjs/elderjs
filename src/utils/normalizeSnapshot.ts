import windowsPathFix from './windowsPathFix';

const normalizeSnapshot = (val) => {
  if (Object.prototype.toString.call(val) === '[object String]') {
    return windowsPathFix(val);
  }
  if (Object.prototype.toString.call(val) === '[object Array]') {
    return val.map(normalizeSnapshot);
  }
  if (Object.prototype.toString.call(val) === '[object Object]') {
    return Object.keys(val).reduce((out, cv) => {
      // eslint-disable-next-line no-param-reassign
      out[cv] = normalizeSnapshot(val[cv]);
      return out;
    }, {});
  }
  return val;
};

export default normalizeSnapshot;
