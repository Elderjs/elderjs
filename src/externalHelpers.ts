/* eslint-disable global-require */
/* eslint-disable import/no-dynamic-require */
import path from 'path';
import fs from 'fs';

import { QueryOptions, SettingsOptions, THelpers } from './utils/types';

let userHelpers;

let cache;

async function externalHelpers({
  settings,
  query,
  helpers,
}: {
  settings: SettingsOptions;
  query: QueryOptions;
  helpers: THelpers;
}) {
  const srcHelpers = path.join(settings.srcDir, 'helpers/index.js');
  try {
    if (!cache) {
      try {
        fs.statSync(srcHelpers);
        const reqHelpers = require(srcHelpers);
        userHelpers = reqHelpers.default || reqHelpers;

        if (typeof userHelpers === 'function') {
          userHelpers = await userHelpers({ settings, query, helpers });
        }
        cache = userHelpers;
      } catch (err) {
        if (err.code === 'ENOENT') {
          if (settings.debug.automagic) {
            console.log(
              `debug.automagic:: We attempted to automatically add in helpers, but we couldn't find the file at ${srcHelpers}.`,
            );
          }
        }
      }
    } else {
      userHelpers = cache;
    }
  } catch (e) {
    console.error(`Error importing ${srcHelpers}`);
    console.error(e);
  }

  return userHelpers;
}

export default externalHelpers;
