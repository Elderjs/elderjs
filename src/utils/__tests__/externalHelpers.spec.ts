import { it, beforeEach, describe, expect, vi } from 'vitest';
import { sep } from 'path';
import normalizeSnapshot from '../normalizeSnapshot.js';
import getConfig from '../getConfig.js';
import externalHelpers from '../externalHelpers';

// process.cwd = () => 'test';

const query = {};

const helpers = {
  inlineSvelteComponent: () => `inlined`,
  permalinks: { key: () => '' },
  shortcode: () => '',
};

describe('#externalHelpers', () => {
  const settings = getConfig({ css: 'inline' });
  beforeEach(() => {
    vi.resetModules();
  });
  it('throws', async () => {
    vi.mock('fs', () => {
      class StatSyncError extends Error {
        code: 'ENOENT';

        constructor(msg: string) {
          super(msg);
          this.code = 'ENOENT';
        }
      }
      return {
        statSync: vi.fn(() => {
          throw new StatSyncError('no file');
        }),
      };
    });
    // eslint-disable-next-line global-require

    expect(await externalHelpers({ settings, query, helpers })).toBeUndefined();
    const modifiedSettings = {
      ...settings,
      srcDir: '',
    };
    expect(
      await externalHelpers({
        settings: modifiedSettings,
        query,
        helpers,
      }),
    ).toBeUndefined();
  });
  it('returns undefined if file is not there', async () => {
    vi.mock('fs', () => ({
      statSync: vi.fn().mockImplementationOnce(() => {
        throw new Error('');
      }),
    }));
    // eslint-disable-next-line global-require

    expect(await externalHelpers({ settings, query, helpers })).toBeUndefined();
  });
  it('works - userHelpers is not a function', async () => {
    vi.mock(
      `src${sep}helpers${sep}index.js`,

      () => ({
        userHelper: () => 'something',
      }),
    );
    vi.mock('fs', () => ({
      statSync: vi.fn().mockImplementationOnce(() => ''),
    }));

    const c1 = await externalHelpers({ settings, query, helpers });
    expect(normalizeSnapshot(c1)).toMatchSnapshot();
    // from cache

    const c2 = await externalHelpers({ settings, query, helpers });
    expect(normalizeSnapshot(c2)).toMatchSnapshot();
  });
  it('works - userHelpers is a function', async () => {
    vi.mock(
      `src${sep}helpers${sep}index.js`,
      () => () =>
        Promise.resolve({
          userHelper: () => 'something',
        }),
    );
    vi.mock('fs', () => ({
      statSync: vi.fn().mockImplementationOnce(() => ''),
    }));
    // eslint-disable-next-line global-require

    const c1 = await externalHelpers({ settings, query, helpers });
    expect(normalizeSnapshot(c1)).toMatchSnapshot();
    // from cache
    const c2 = await externalHelpers({ settings, query, helpers });
    expect(normalizeSnapshot(c2)).toMatchSnapshot();
  });
});
