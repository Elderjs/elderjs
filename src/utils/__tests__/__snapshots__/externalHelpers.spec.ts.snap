// Vitest Snapshot v1

exports[`#externalHelpers > works - userHelpers is a function 1`] = `undefined`;

exports[`#externalHelpers > works - userHelpers is a function 2`] = `undefined`;

exports[`#externalHelpers > works - userHelpers is not a function 1`] = `undefined`;

exports[`#externalHelpers > works - userHelpers is not a function 2`] = `undefined`;
